### ISSUE TITLE
RFC: JavaScript Code Style

### ISSUE THREAD

**Issue by invenio-developers** [inveniosoftware/invenio#1769]
<a href="https://github.com/jirikuncar"><img src="https://avatars.githubusercontent.com/u/189798?" align="left" width="96" height="96" hspace="10"></img></a> **Issue by [jirikuncar](https://github.com/jirikuncar)**
_Thursday May 22, 2014 at 11:54 GMT_
_Originally opened as https://github.com/jirikuncar/invenio/issues/262_

---

| Revision | Date | Author |
| --- | --- | --- |
| Initial | 2014-05-22 | @jirikuncar |
# Idea

It'd be nice to use the standard way for JavaScript.
# Proposals
- @lnielsen-cern: See previous discussion as well http://invenio-software.org/wiki/Talk/JSFrameworks
- @giokokos https://github.com/airbnb/javascript
### Testing
- @greut The first point to enforce regarding JS is **no inline JavaScript, no `onclick`, no `href=javascript:`**. The second would be **tests**, [QUnit](http://qunitjs.com/) is super simple and nice. 
  Extra pain points: [jslint](http://jslint.com/), [jsfmt](http://rdio.github.io/jsfmt/)
### Related
- Require.js #1794, #1830, #1833

**Comment by jacquerie** [comment]
It appears that this RFC didn't reach consensus, so no decision was made. Let me try to rekindle the discussion with some organization and some proposals.

The discussion started from code style, then touched code quality, the use of frameworks, testing, and documentation. I think these are very different problems, which require different discussions.

1. JavaScript code style: IMHO, not as important as the others, but still nice to have. Let me mention [JSCS](https://github.com/jscs-dev/node-jscs), a configurable tool to enforce code style conventions. Once we decide on a style guide (or perhaps we can lean on the defaults of some project we like), it should make the job much easier.
2. JavaScript code quality: I think there are two parts to this. First we have the style of JavaScript files: a tool that was already mentioned is [JSHint](https://github.com/jshint/jshint), which will require a discussion on the settings. Then there's the problem of the obtrusiveness of the JavaScript in A LOT of the templates. Judging from the output of `ag onclick` this will require A LOT of work.
3. Use of JavaScript frameworks: consensus in #2110 is that this shouldn't be pursued.
4. Testing JavaScript libraries: RFC #2221 is still open, but, from what I've seen, Jasmine is the de-facto standard.
5. Documenting JavaScript: RFC #2845 is still open, and this is where it should be discussed.

Related to these, and in particular to the first two, is the question of having some kind of task runner for JavaScript tasks, like Grunt or Gulp. Should I open an RFC for this?

In summary I propose to:

1. Use this RFC to discuss JSCS's settings
2. Start a new RFC about JSHint's settings
3. Start an effort to refactor the templates
4. Use the existing RFCs to discuss testing libraries and documentation solutions
5. Start a new RFC about a JavaScript task runner

**Comment by jacquerie** [comment]
To follow up on yesterday's question, no, JSCS has no "autoformatting" functionality, but it's been actively worked on, so it should be here soon™: https://github.com/jscs-dev/node-jscs/issues/516.

**Comment by hzoo** [comment]
JSCS just added the `--autofix` option as of [1.12](https://github.com/jscs-dev/node-jscs/releases/tag/v1.12.0) (on most of the whitespace rules).

You guys can start with using a preset (I see airbnb was mentioned in the OP) and then nulling out any rules you don't think or adding more that you want. Also feel free to create issues/PRs for new rules/options that could be added in!

**Comment by jacquerie** [comment]
So, let's get back to our discussion. I'll take the following code in Invenio (from https://github.com/inveniosoftware/invenio/blob/ad39db02238330f419bb73d40d708df059d05cf8/invenio/base/static/js/invenio.js#L43-L59) and rewrite it using 3 different presets: AirBnB, Crockford, jQuery.

```js
$('[data-menu="click"]').each(function (index) {
  var that = this,
    $menu = $(this).parent(),
    $li = $menu.parent(),
    message_update = function(obj) {
      $(that).unbind('click');
      $.ajax({
        url: $(that).attr('data-menu-source')
      }).done(function (data) {
        $menu.parent().append(data);
      });
    };
  $(that).on('click', message_update);
  $li.addClass('dropdown');
  $menu.attr('data-toggle', "dropdown");
  $menu.dropdown();
});
```

### 1. AirBnB

```js
$('[data-menu="click"]').each(function(index) {
  var _this = this;
  var $menu = $(this).parent();
  var $li = $menu.parent();
  var messageUpdate = function(obj) {
    $(_this).unbind('click');
    $.ajax({
      url: $(_this).attr('data-menu-source')
    }).done(function(data) {
      $menu.parent().append(data);
    });
  };

  $(_this).on('click', messageUpdate);
  $li.addClass('dropdown');
  $menu.attr('data-toggle', 'dropdown');
  $menu.dropdown();
});
```

### 2. Crockford

```js
$('[data-menu="click"]').each(function (index) {
    var that = this,
        $menu = $(this).parent(),
        $li = $menu.parent(),
        message_update = function (obj) {
            $(that).unbind('click');
            $.ajax({
                url: $(that).attr('data-menu-source')
            }).done(function (data) {
                $menu.parent().append(data);
            });
        };
    $(that).on('click', message_update);
    $li.addClass('dropdown');
    $menu.attr('data-toggle', "dropdown");
    $menu.dropdown();
});
```

### 3. jQuery

```js
$( '[data-menu="click"]' ).each( function( index ) {
  var that = this,
    $menu = $( this ).parent(),
    $li = $menu.parent(),
    messageUpdate = function( obj ) {
      $( that ).unbind( "click" );
      $.ajax( {
        url: $( that ).attr( "data-menu-source" )
      } ).done( function( data ) {
        $menu.parent().append( data );
      } );
    };
  $( that ).on( "click", messageUpdate );
  $li.addClass( "dropdown" );
  $menu.attr( "data-toggle", "dropdown" );
  $menu.dropdown();
} );
```

Personally, I'd go with number 1.

Here's also an estimate of the effort needed (obtained running `jscs -p PRESET invenio/modules/*/static`):
1. AirBnB: 20231 code style errors found.
2. Crockford: 22815 code style errors found.
3. jQuery: 33699 code style errors found.

**Comment by jirikuncar** [comment]
I would like 2. Crockford with 2 space indentation but I'm also fine with 1. AirBnB with possibility to join `var` definitions.

```javascript
var foo = 1,
    bar = 'baz',
    fb = function() {
      return 1;
    };
```

**Comment by jmartinm** [comment]
1 seems good to me. Agree with Jiri that I would allow join var definitions (as long as they are properly aligned with each other)

**Comment by bouzlibop** [comment]
I give my vote to 1. AriBnB and as mentioned above - with the possibility to have var definitions joined.

BTW, in regards to proposal 5:
> Start a new RFC about a JavaScript task runner

What use cases do you see @jacquerie for js task runner in `Invenio` in this moment ?

**Comment by jacquerie** [comment]
As a wrapper for the tooling we introduce in this and several other RFCs, for programmer convenience. For example, `grunt jscs` might check the code style in all Invenio, using the right paths and excluding files which we don't need to check, without having to remember a long `jscs` command line.

**Comment by lnielsen** [comment]
It's highly preferable (in my opinion a must) that any linting is integrated with inveniosoftware/kwalitee so that there's only one tool for checking Invenio source code. 1) It's easier for developers (just one tool to run on your commits) 2) Kwalitee is integrated with GitHub so PR's can be checked automatically.

Grunt could potentially be integrated with Kwalitee, so just wanted to emphasise that integration with kwalitee has to be thought of when taking a decision.

**Comment by bouzlibop** [comment]
I would propose to extend `inveniomanage` and `kwalitee` to provide necessary aliases/pipelining/quality checks and avoid adding new tools (`grunt`).

**Comment by hzoo** [comment]
You could also use npm to hold scripts if not grunt/gulp. So in package.json (example from jscs itself)

```js
"scripts": {
    "lint": "jshint . && node bin/jscs lib test bin publish"
}
```

**Comment by jacquerie** [comment]
:+1: Ok, I think the consensus is against having a task runner, so I'll drop that proposal.

**Comment by jalavik** [comment]
@jacquerie We could add it as a suggestion to the invenio kwalitee tool as @lnielsen suggested. Sounds like a nice compromise.
